File: ktorOld/main/kotlin/services/FileInitializationService.kt
package services

import com.google.cloud.storage.Storage
import config.GcsConfig
import schema.model.*

class FileInitializationService(
    private val storageService: StorageService,
    private val storageDbService: StorageDbService,
    private val storage: Storage,
    private val gcsConfig: GcsConfig
) {
    /**
     * Initialize required directories and upload demo files
     */
    fun initializeFileSystem() {
        println("🗂️  Initializing file system...")

        // Create required directories
        createRequiredDirectories()

        // Upload demo template files
        uploadDemoTemplateFiles()

        println("✅ File system initialization completed")
    }

    /**
     * Create required directories in the storage system
     */
    private fun createRequiredDirectories() {
        val requiredDirectories = listOf(
            CreateFolderInput(path = "public", protected = true),
            CreateFolderInput(path = "public/templates", protected = true),
            CreateFolderInput(path = "public/templates/covers", protected = true)
        )

        for (input in requiredDirectories) {
            createDirectoryIfNotExists(input)
        }
    }

    /**
     * Create a directory if it doesn't exist
     */
    private fun createDirectoryIfNotExists(input: CreateFolderInput) {
        // Check if directory exists in the bucket
        val bucketPath = "${input.path}/"
        try {
            val blob = storage.get(gcsConfig.bucketName, bucketPath)
            if (blob == null) {
                // Create the directory marker in the bucket
                val blobInfo = com.google.cloud.storage.BlobInfo.newBuilder(gcsConfig.bucketName, bucketPath)
                    .setContentType("application/x-directory")
                    .build()
                storage.create(blobInfo, ByteArray(0))
                println("   📁 Created directory in bucket: ${input.path}")
            } else {
                println("   📁 Directory already exists in bucket: ${input.path}")
            }
        } catch (e: Exception) {
            println("   ⚠️  Warning: Could not create directory in bucket: ${input.path} - ${e.message}")
        }

        val result = storageService.createFolder(input)
        if (result.success) {
            println("   📁 Created directory with permissions: ${input.path}")
        } else {
            println("   ⚠️  Warning: Could not register directory permissions in database: ${result.message}")
        }
    }

    /**
     * Upload demo template files from resources
     */
    private fun uploadDemoTemplateFiles() {
        val demoFiles = listOf(
            "demo1.jpg",
            "demo2.jpg",
            "demo3.jpg",
            "demo4.jpg"
        )

        for (fileName in demoFiles) {
            uploadDemoFileIfNotExists(fileName)
        }

        // After processing all files, verify they are in the database
        println("   🔍 Verifying demo files are registered in database...")
        verifyDemoFilesInDatabase()
    }

    /**
     * Upload a demo file if it doesn't exist in storage
     */
    private fun uploadDemoFileIfNotExists(fileName: String) {
        val bucketPath = "public/templates/covers/$fileName"

        // Check if file exists in the bucket
        try {
            val blob = storage.get(gcsConfig.bucketName, bucketPath)
            if (blob != null && blob.size > 0) {
                println("   🖼️  Demo file already exists in bucket: $fileName")
                return
            }
        } catch (e: Exception) {
            println("   ⚠️  Could not check bucket for file: $fileName - ${e.message}")
        }

        // Upload file from resources using StorageService
        try {
            val resourcePath = "/img/$fileName"
            val inputStream = this::class.java.getResourceAsStream(resourcePath)

            if (inputStream == null) {
                println("   ❌ Resource not found: $resourcePath")
                return
            }

            inputStream.use { stream ->
                val contentType = getContentTypeFromFileName(fileName)

                val result = storageService.uploadFile(
                    path = bucketPath,
                    inputStream = stream,
                    contentType = contentType,
                    originalFilename = fileName,
                    location = UploadLocation.TEMPLATE_COVER
                )

                if (result.success) {
                    println("   🖼️  Uploaded demo file: $fileName")
                } else {
                    println("   ❌ Failed to upload demo file $fileName: ${result.message}")
                }
            }
        } catch (e: Exception) {
            println("   ❌ Failed to upload demo file $fileName: ${e.message}")
        }
    }

    /**
     * Verify that all demo files are properly accessible in the bucket
     */
    private fun verifyDemoFilesInDatabase() {
        val demoFiles = listOf(
            "demo1.jpg" to "public/templates/covers/demo1.jpg",
            "demo2.jpg" to "public/templates/covers/demo2.jpg",
            "demo3.jpg" to "public/templates/covers/demo3.jpg",
            "demo4.jpg" to "public/templates/covers/demo4.jpg"
        )

        for ((fileName, bucketPath) in demoFiles) {
            try {
                val blob = storage.get(gcsConfig.bucketName, bucketPath)
                if (blob != null && blob.size > 0) {
                    println("   ✅ Demo file exists in bucket: $fileName")
                } else {
                    println("   ❌ Demo file NOT found in bucket: $fileName")
                }
            } catch (e: Exception) {
                println("   ❌ Error checking demo file $fileName: ${e.message}")
            }
        }
    }

    /**
     * Get demo files that are available for templates
     * Since files are no longer automatically stored in DB,
     * we'll register them in DB when they're first used
     */
    suspend fun getDemoFileIds(): List<Long> {
        val demoFiles = listOf(
            "public/templates/covers/demo1.jpg",
            "public/templates/covers/demo2.jpg",
            "public/templates/covers/demo3.jpg",
            "public/templates/covers/demo4.jpg"
        )

        println("   🔍 Checking for demo files and registering as needed...")
        val fileIds = demoFiles.mapNotNull { path ->
            // First check if already in DB
            val existingFile = storageDbService.fileByPath(path)
            if (existingFile != null) {
                println("   ✅ Found demo file in database: $path (ID: ${existingFile.id})")
                existingFile.id
            } else {
                // Check if exists in bucket and register it for use
                try {
                    val blob = storage.get(gcsConfig.bucketName, path)
                    if (blob != null && blob.size > 0) {
                        val fileName = path.substringAfterLast('/')
                        val directoryPath = path.substringBeforeLast('/')
                        val contentType = getContentTypeFromFileName(fileName)

                        val fileEntity = storageDbService.createFile(
                            path = path,
                            name = fileName,
                            directoryPath = directoryPath,
                            size = blob.size,
                            contentType = contentType.value,
                            md5Hash = blob.md5ToHexString
                        )

                        println("   ✅ Registered demo file for use: $path (ID: ${fileEntity.id})")
                        fileEntity.id
                    } else {
                        println("   ❌ Demo file NOT found in bucket: $path")
                        null
                    }
                } catch (e: Exception) {
                    println("   ❌ Error accessing demo file $path: ${e.message}")
                    null
                }
            }
        }

        println("   📊 Total demo files available: ${fileIds.size} out of ${demoFiles.size}")
        return fileIds.map { it.value }
    }

    /**
     * Register file usage for template covers
     */
    suspend fun registerTemplateFileUsage(templateId: Int, fileId: Long) {
        val fileEntity = storageDbService.fileById(fileId)
        if (fileEntity != null) {
            val input = RegisterFileUsageInput(
                filePath = fileEntity.path,
                usageType = "template_cover",
                referenceId = templateId.toLong(),
                referenceTable = "templates"
            )

            val result = storageDbService.registerFileUsage(input)
            if (result.success) {
                println("   📎 Registered file usage for template $templateId with file $fileId at path: ${fileEntity.path}")
            } else {
                println("   ⚠️  Failed to register file usage: ${result.message}")
            }
        } else {
            println("   ⚠️  File with ID $fileId not found")
        }
    }

    private fun getContentTypeFromFileName(fileName: String): ContentType {
        return when (fileName.substringAfterLast(".").lowercase()) {
            "jpg", "jpeg" -> ContentType.JPEG
            "png" -> ContentType.PNG
            "gif" -> ContentType.GIF
            "svg" -> ContentType.SVG
            else -> ContentType.JPEG
        }
    }
}

File: server/storage/demo/fileInitializationService.ts
import logger from "@/server/lib/logger";
import * as fs from "fs/promises";
import * as path from "path";
import { getStorageService, type StorageService } from "../storage.service";
import * as Types from "@/server/types";

/**
 * Interface for FileInitializationService
 */
export interface IFileInitializationService {
  /**
   * Initialize required directories and upload demo files
   */
  initializeFileSystem(): Promise<void>;

  /**
   * Create required directories in the storage system
   */
  createRequiredDirectories(): Promise<void>;

  /**
   * Create a directory if it doesn't exist
   */
  createDirectoryIfNotExists(input: Types.FolderCreateInput): Promise<void>;

  /**
   * Upload demo template files from resources
   */
  uploadDemoTemplateFiles(): Promise<void>;

  /**
   * Upload a demo file if it doesn't exist in storage
   */
  uploadDemoFileIfNotExists(fileName: string): Promise<void>;

  /**
   * Verify that all demo files are properly accessible in the bucket
   */
  verifyDemoFilesInDatabase(): Promise<void>;

  /**
   * Get demo files that are available for templates
   * Files are automatically registered in the database when uploaded via StorageService
   * This method retrieves their database IDs for use in templates
   */
  getDemoFileIds(): Promise<bigint[]>;

  /**
   * Register file usage for template covers
   */
  registerTemplateFileUsage(templateId: number, fileId: bigint): Promise<void>;

  /**
   * Get content type from file name
   */
  getContentTypeFromFileName(fileName: string): string;
}

export class FileInitializationService implements IFileInitializationService {
  private readonly storageService: StorageService;

  constructor(storageService: StorageService) {
    this.storageService = storageService;
  }

  /**
   * Create a directory if it doesn't exist
   */
  async createDirectoryIfNotExists(input: Types.FolderCreateInput): Promise<void> {
    try {
      // Check if directory exists in storage
      const directoryInfo = await this.storageService.directoryInfoByPath(input.path);

      if (directoryInfo) {
        logger.info(`   📁 Directory already exists: ${input.path}`);
        return;
      }

      // Create the directory
      const result = await this.storageService.createFolder(input);

      if (result.success) {
        logger.info(`   📁 Created directory: ${input.path}`);
      } else {
        logger.warn(`   ⚠️  Warning: Could not create directory: ${input.path} - ${result.message}`);
      }
    } catch (error) {
      logger.error(`   ❌ Error creating directory ${input.path}:`, error);
      throw error;
    }
  }

  // Placeholder implementations for other methods
  async initializeFileSystem(): Promise<void> {
    logger.info("🗂️  Initializing file system...");

    // Create required directories
    await this.createRequiredDirectories();

    // Upload demo template files
    await this.uploadDemoTemplateFiles();

    logger.info("✅ File system initialization completed");
  }

  async createRequiredDirectories(): Promise<void> {
    const requiredDirectories: Types.FolderCreateInput[] = [
      { path: "public", protected: true },
      { path: "public/templates", protected: true },
      { path: "public/templates/covers", protected: true },
    ];

    for (const input of requiredDirectories) {
      await this.createDirectoryIfNotExists(input);
    }
  }

  async uploadDemoTemplateFiles(): Promise<void> {
    const demoFiles = ["demo1.jpg", "demo2.jpg", "demo3.jpg", "demo4.jpg"];

    for (const fileName of demoFiles) {
      await this.uploadDemoFileIfNotExists(fileName);
    }

    // After processing all files, verify they are in the database
    logger.info("   🔍 Verifying demo files are registered in database...");
    await this.verifyDemoFilesInDatabase();
  }

  async uploadDemoFileIfNotExists(fileName: string): Promise<void> {
    const bucketPath = `public/templates/covers/${fileName}`;

    try {
      // Check if file exists in storage
      const fileInfo = await this.storageService.fileInfoByPath(bucketPath);
      if (fileInfo) {
        logger.info(`   🖼️  Demo file already exists in bucket: ${fileName}`);
        return;
      }
    } catch (error) {
      logger.warn(`   ⚠️  Could not check bucket for file: ${fileName} - ${error}`);
    }

    // Upload file from resources using StorageService
    try {
      const resourcePath = path.join(__dirname, "res", fileName);

      // Read file from disk
      const fileBuffer = await fs.readFile(resourcePath);
      const contentType = this.getContentTypeFromFileName(fileName);

      const result = await this.storageService.uploadFile(bucketPath, contentType, fileBuffer);

      if (result.success) {
        logger.info(`   🖼️  Uploaded demo file: ${fileName}`);
      } else {
        logger.error(`   ❌ Failed to upload demo file ${fileName}: ${result.message}`);
      }
    } catch (error) {
      if (error instanceof Error && "code" in error && error.code === "ENOENT") {
        logger.error(`   ❌ Resource not found: ${fileName}`);
      } else {
        logger.error(`   ❌ Failed to upload demo file ${fileName}:`, error);
      }
    }
  }

  async verifyDemoFilesInDatabase(): Promise<void> {
    const demoFiles = [
      {
        fileName: "demo1.jpg",
        bucketPath: "public/templates/covers/demo1.jpg",
      },
      {
        fileName: "demo2.jpg",
        bucketPath: "public/templates/covers/demo2.jpg",
      },
      {
        fileName: "demo3.jpg",
        bucketPath: "public/templates/covers/demo3.jpg",
      },
      {
        fileName: "demo4.jpg",
        bucketPath: "public/templates/covers/demo4.jpg",
      },
    ];

    for (const { fileName, bucketPath } of demoFiles) {
      try {
        const fileInfo = await this.storageService.fileInfoByPath(bucketPath);
        if (fileInfo) {
          logger.info(`   ✅ Demo file exists in bucket: ${fileName}`);
        } else {
          logger.error(`   ❌ Demo file NOT found in bucket: ${fileName}`);
        }
      } catch (error) {
        logger.error(`   ❌ Error checking demo file ${fileName}:`, error);
      }
    }
  }

  async getDemoFileIds(): Promise<bigint[]> {
    const demoFiles = [
      "public/templates/covers/demo1.jpg",
      "public/templates/covers/demo2.jpg",
      "public/templates/covers/demo3.jpg",
      "public/templates/covers/demo4.jpg",
    ];

    logger.info("   🔍 Checking for demo files and registering as needed...");

    const fileIds: bigint[] = [];
    const { StorageDbRepository } = await import("@/server/db/repo");

    for (const filePath of demoFiles) {
      try {
        // First check if already in DB
        const existingFile = await StorageDbRepository.fileByPath(filePath);

        if (existingFile) {
          logger.info(`   ✅ Found demo file in database: ${filePath} (ID: ${existingFile.id})`);
          fileIds.push(existingFile.id);
          continue;
        }

        // Check if exists in bucket and register it for use
        const fileInfo = await this.storageService.fileInfoByPath(filePath);

        if (fileInfo) {
          // File exists in bucket but not in database - register it
          const fileEntity = await StorageDbRepository.createFile(
            filePath,
            true // isProtected = true for demo files
          );

          logger.info(`   ✅ Registered demo file for use: ${filePath} (ID: ${fileEntity.id})`);
          fileIds.push(fileEntity.id);
        } else {
          logger.error(`   ❌ Demo file NOT found in bucket: ${filePath}`);
        }
      } catch (error) {
        logger.error(`   ❌ Error accessing demo file ${filePath}:`, error);
      }
    }

    logger.info(`   📊 Total demo files available: ${fileIds.length} out of ${demoFiles.length}`);
    return fileIds;
  }

  async registerTemplateFileUsage(templateId: number, fileId: bigint): Promise<void> {
    try {
      const fileInfo = await this.storageService.fileInfoByDbFileId(fileId);

      if (fileInfo) {
        // NOTE: File usage registration through StorageService.registerFileUsage()
        // is not yet implemented in the StorageService interface.
        // Currently, file usage tracking happens when templates reference the files
        // in the database through foreign key relationships.
        //
        // Expected future behavior (to match Kotlin implementation):
        // - Register file usage with type "template_cover"
        // - Track reference to template ID and "templates" table
        // - This enables checking if files can be safely deleted
        logger.info(`   📎 File ${fileId} (${fileInfo.path}) will be used for template ${templateId}`);
        logger.info(`   ℹ️  File usage tracking happens through template-file relationships`);
      } else {
        logger.warn(`   ⚠️  File with ID ${fileId} not found`);
      }
    } catch (error) {
      logger.error(`   ❌ Error accessing file ${fileId}:`, error);
    }
  }

  getContentTypeFromFileName(fileName: string): string {
    const extension = fileName.substring(fileName.lastIndexOf(".") + 1).toLowerCase();

    switch (extension) {
      case "jpg":
      case "jpeg":
        return "image/jpeg";
      case "png":
        return "image/png";
      case "gif":
        return "image/gif";
      case "webp":
        return "image/webp";
      case "pdf":
        return "application/pdf";
      case "doc":
        return "application/msword";
      case "docx":
        return "application/vnd.openxmlformats-officedocument.wordprocessingml.document";
      case "xls":
        return "application/vnd.ms-excel";
      case "xlsx":
        return "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";
      case "txt":
        return "text/plain";
      case "zip":
        return "application/zip";
      case "rar":
        return "application/vnd.rar";
      case "mp4":
        return "video/mp4";
      case "mp3":
        return "audio/mpeg";
      case "wav":
        return "audio/wav";
      case "otf":
        return "font/otf";
      case "ttf":
        return "font/ttf";
      case "woff":
        return "font/woff";
      case "woff2":
        return "font/woff2";
      default:
        return "image/jpeg";
    }
  }
}

/**
 * Factory function to create FileInitializationService instance
 */
export const createFileInitializationService = async (): Promise<FileInitializationService> => {
  const storageService = await getStorageService();
  return new FileInitializationService(storageService);
};
